File: day07.py
import time

start = time.time()

infile = open('inputs/day07.txt', 'r')
lines = infile.readlines()
infile.close()

lines = [line.strip().split(':') for line in lines]
lines = [
    {
        'total': int(line[0]), 
        'operands': [int(num) for num in line[1].strip().split(' ')]
    } for line in lines
]

def is_possible(target, operands, allow_third_operator=False):
    if len(operands) == 1:
        return target == operands[0]
    plus = is_possible(target, [operands[0]+operands[1]] + operands[2:], allow_third_operator)
    if plus: return True
    times = is_possible(target, [operands[0]*operands[1]] + operands[2:], allow_third_operator) 
    if times: return True
    if allow_third_operator:
        concat = is_possible(target, [int(str(operands[0]) + str(operands[1]))] + operands[2:], allow_third_operator)
        if concat: return True
    return False


total = 0
for line in lines:
    if is_possible(line['total'], line['operands']):
        total += line['total']
print(f'Part 1: {total}')

total = 0
for line in lines:
    if is_possible(line['total'], line['operands'], allow_third_operator=True):
        total += line['total']

print(f'Part 2: {total}')

end = time.time()

print(f'time: {end - start}s')

File: day07.js
console.time('time');

const fs = require("fs");

const lines = fs
  .readFileSync("inputs/day07.txt", {encoding: 'utf-8'})
  .split("\n")
  .map((line) => line.split(": "))
  .map((line) => {
    return {
      total: Number(line[0]),
      operands: line[1].split(" ").map((num) => Number(num)),
    };
  });

let total = 0;
for (let line of lines) {
  if (isPossible(line.total, line.operands)) total += line.total;
}
console.log(`Part 1: ${total}`);

total = 0;
for (let line of lines) {
  if (isPossible(line.total, line.operands, true)) {
    total += line.total;
  }
}
console.log(`Part 2: ${total}`);

function isPossible(target, operands, allowThirdOperator = false) {
  if (operands.length === 1) return target === operands[0];
  const plus = isPossible(
    target,
    [operands[0] + operands[1]].concat(operands.slice(2)),
    allowThirdOperator
  );
  if (plus) return true;
  const times = isPossible(
    target,
    [operands[0] * operands[1]].concat(operands.slice(2)),
    allowThirdOperator
  );
  if (times) return true;
  if (allowThirdOperator) {
    const conc = isPossible(
      target,
      [Number(operands[0].toString() + operands[1].toString())].concat(
        operands.slice(2)
      ),
      allowThirdOperator
    );
    if (conc) return true;
  }
  return false;
}

console.timeEnd('time')